File: tests/test_gmdh.py
"""
All tests for gmdh module.
"""

#pylint: disable=invalid-name

import pytest
import numpy as np
import gmdh

@pytest.fixture
def criterions_classes():
    """Returns all criterion classes of gmdh module"""
    return [gmdh.Criterion, gmdh.ParallelCriterion, gmdh.SequentialCriterion]

@pytest.fixture
def combined_criterions_classes():
    """Returns only combined criterion classes of gmdh module"""
    return [gmdh.ParallelCriterion, gmdh.SequentialCriterion]

class TestDataPreparations:
    """
    Class containing methods for testing gmdh module functions
    that prepare user data for further use in gmdh models.
    """

    @pytest.mark.parametrize(
        'original_data, ts_lags',
        [([1, 2, 3, 4, 5, 6], 1),
        ([1, 2, 3, 4, 5, 6], 4),
        ([1, 2, 3, 4, 5, 6], 5),
        ([np.random.random() for i in range(10000)], 50)])
    def test_time_series_transformation(self, original_data, ts_lags):
        """
        Testing gmdh.time_series_transformation() method using both small and large original_data.
        All values of the input argument are correct.
        """
        X, y = gmdh.time_series_transformation(original_data, lags=ts_lags)
        for i in range(len(original_data) - ts_lags):
            assert np.array_equal(X[i], original_data[i:i+ts_lags])
        assert np.array_equal(y, original_data[ts_lags:])

    @pytest.mark.parametrize(
        'original_data, ts_lags',
        [([1, 2, 3, 4, 5, 6], 0),  # lags = 0
        ([1, 2, 3, 4, 5, 6], -1),  # lags < 0
        ([1, 2, 3, 4, 5, 6], 6),  # lags = time series size
        ([1, 2, 3, 4, 5, 6], 7),  # lags > time series size
        ([], 2)])  # empty time series
    def test_time_series_transformation_value_error(self, original_data, ts_lags):
        """
        Testing gmdh.time_series_transformation() method using incorrect input arguments values.
        Expected result is ValueError.
        """
        with pytest.raises(ValueError):
            gmdh.time_series_transformation(original_data, lags=ts_lags)

    @pytest.mark.parametrize(
        'original_data, ts_lags',
        [([1, 2, 3, 4, 5, 6], 3.5),  # not integer lags
        ([1, 2, 3, 4, 5, 6], 'a'),  # not integer lags
        (['a', 'b', 'c', 'd'], 3)])  # time series data contains strings
    def test_time_series_transformation_type_error(self, original_data, ts_lags):
        """
        Testing gmdh.time_series_transformation() method using incorrect input arguments types.
        Expected result is TypeError.
        """
        with pytest.raises(TypeError):
            gmdh.time_series_transformation(original_data, lags=ts_lags)

    def test_split_data(self):
        """
        Testing gmdh.split_data() method using simple X and y arrays without shuffling.
        """
        X = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
        y = [10, 11, 12]
        x_train, x_test, y_train, y_test = gmdh.split_data(X, y, test_size=0.33)
        assert np.array_equal(x_train, [[1, 2, 3], [4, 5, 6]])
        assert np.array_equal(x_test, [[7, 8, 9]])
        assert np.array_equal(y_train, [10, 11])
        assert np.array_equal(y_test, [12])

    def test_split_data_with_shuffle(self):
        """
        Testing gmdh.split_data() method using simple X and y arrays
        with shuffling and the same random_state param 2 times.
        The expected resulting arrays for the first and second time should be equal.
        """
        X = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
        y = [10, 11, 12]
        x_train, x_test, y_train, y_test = \
            gmdh.split_data(X, y, test_size=0.66, shuffle=True, random_state=1)
        x_train2, x_test2, y_train2, y_test2 = \
            gmdh.split_data(X, y, test_size=0.66, shuffle=True, random_state=1)
        assert x_train.shape == (1, 3) and np.array_equal(x_train, x_train2)
        assert x_test.shape == (2, 3) and np.array_equal(x_test, x_test2)
        assert y_train.shape == (1,) and np.array_equal(y_train, y_train2)
        assert y_test.shape == (2,) and np.array_equal(y_test, y_test2)

    @pytest.mark.parametrize(
        'X, y, test_size',
        [([], [], 0.2),  # empty arrays
        ([[1, 2, 3], [4, 5, 6]], [7, 8], 0.1),  # empty test arrays because of small test_size
        ([[1, 2, 3], [4, 5, 6]], [7, 8], 0.9),  # empty train arrays because of large test_size
        ([[1, 2, 3], [4, 5, 6]], [7], 0.5),  # different X rows and y size values
        ([[1, 2, 3], [4, 5, 6]], [7, 8], 0),  # test_size = 0
        ([[1, 2, 3], [4, 5, 6]], [7, 8], -1),  # test_size < 0
        ([[1, 2, 3], [4, 5, 6]], [7, 8], 1),  # test_size = 1
        ([[1, 2, 3], [4, 5, 6]], [7, 8], 10)])  # test_size > 1
    def test_split_data_value_error(self, X, y, test_size):
        """
        Testing gmdh.split_data() method using incorrect input arguments values.
        Expected result is ValueError.
        """
        with pytest.raises(ValueError):
            gmdh.split_data(X, y, test_size=test_size)

class TestCriterions:
    """
    Class containing methods for testing gmdh module criterions.
    """
    # pylint: disable=redefined-outer-name

    @pytest.mark.parametrize(
        'criterion_type',
        [1, gmdh.Solver.ACCURATE])
    def test_invalid_criterion_type(self, criterion_type, criterions_classes):
        """
        Testing all criterion classes using incorrect types of `criterion_type` argument.
        Expected result is TypeError.
        """
        for criterion_class in criterions_classes:
            # invalid initialization
            with pytest.raises(TypeError):
                criterion_class(criterion_type=criterion_type)

            # invalid assignment
            criterion = criterion_class()
            with pytest.raises(TypeError):
                criterion.criterion_type=criterion_type

    @pytest.mark.parametrize(
        'solver',
        [1, gmdh.CriterionType.REGULARITY])
    def test_invalid_solver(self, solver, criterions_classes):
        """
        Testing all criterion classes using incorrect types of `solver` argument.
        Expected result is TypeError.
        """
        for criterion_class in criterions_classes:
            # invalid initialization
            with pytest.raises(TypeError):
                criterion_class(solver=solver)

            # invalid assignment
            criterion = criterion_class()
            with pytest.raises(TypeError):
                criterion.solver=solver

    @pytest.mark.parametrize(
        'alpha',
        ['a', None, [1, 2, 3], gmdh.Solver.FAST])
    def test_alpha_type_error(self, alpha):
        """
        Testing `ParallelCriterion` class using incorrect types of `alpha` argument.
        Expected result is TypeError.
        """
        # invalid initialization
        with pytest.raises(TypeError):
            gmdh.ParallelCriterion(alpha=alpha)

        # invalid assignment
        criterion = gmdh.ParallelCriterion()
        with pytest.raises(TypeError):
            criterion.alpha=alpha

    @pytest.mark.parametrize(
        'alpha',
        [0, 1, 2.5, -0.5])
    def test_alpha_value_error(self, alpha):
        """
        Testing `ParallelCriterion` class using incorrect values of `alpha` argument.
        Expected result is ValueError.
        """
        # invalid initialization
        with pytest.raises(ValueError):
            gmdh.ParallelCriterion(alpha=alpha)

        # invalid assignment
        criterion = gmdh.ParallelCriterion()
        with pytest.raises(ValueError):
            criterion.alpha=alpha

    @pytest.mark.parametrize(
        'second_criterion_type',
        [1, gmdh.Solver.ACCURATE])
    def test_invalid_second_criterion_type(self, second_criterion_type, \
        combined_criterions_classes):
        """
        Testing combined criterion classes using incorrect types
        of `second_criterion_type` argument.
        Expected result is TypeError.
        """
        for criterion_class in combined_criterions_classes:
            # invalid initialization
            with pytest.raises(TypeError):
                criterion_class(second_criterion_type=second_criterion_type)

            # invalid assignment
            criterion = criterion_class()
            with pytest.raises(TypeError):
                criterion.second_criterion_type=second_criterion_type

    def test_combined_criterions_in_fit(self, combined_criterions_classes):
        """
        Testing combined criterion classes by using them in fitting Combi model.
        """
        X, y = gmdh.time_series_transformation([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11], lags=3)
        x_train, x_test, y_train, y_test = gmdh.split_data(X, y)
        model = gmdh.Combi()
        for criterion_class in combined_criterions_classes:
            model.fit(x_train, y_train, criterion_class())
            y_pred = model.predict(x_test)
            assert np.array_equal(y_test, y_pred)

File: tests/test_gmdh.cpp
#include "test_setup.h"

bool TestGmdhModel::SKIP_FIXTURES = false;

TEST(testTimeSeriesTransformation, testCorrectData) {
    std::pair<VectorXd, int> origData = { VectorXd{{ 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 }}, 4 };
    GTEST_EXPECT_NO_DEATH({
        auto timeSeries = timeSeriesTransformation(origData.first, origData.second);
        
        for (auto i = 0; i != origData.first.size() - origData.second; ++i) {
            for (auto j = 0; j != origData.second; ++j)
                ASSERT_EQ(origData.first[j + i], timeSeries.first.row(i)[j]);
            ASSERT_EQ(origData.first[i + origData.second], timeSeries.second[i]);
        }
    }) << "[ TEST_MSG ]: death timeSeriesTransformation";
    if (HasFailure())
        TestGmdhModel::SKIP_FIXTURES = true;
}

TEST(testTimeSeriesTransformation, testIncorrectData) {
    std::pair<VectorXd, int> origData = { VectorXd{{ 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 }}, 4 };
    VectorXd errorData;
    EXPECT_THROW(timeSeriesTransformation(errorData, 1), std::invalid_argument) << "[ TEST_MSG ]: empty timeSeries VectorXd in argument"; 
    EXPECT_THROW(timeSeriesTransformation(origData.first, -1), std::invalid_argument) << "[ TEST_MSG ]: negative number of lags";
    EXPECT_THROW(timeSeriesTransformation(origData.first, 0), std::invalid_argument) << "[ TEST_MSG ]: zero number of lags";
    EXPECT_THROW(timeSeriesTransformation(origData.first, origData.first.size()), std::invalid_argument) << "[ TEST_MSG ]: number of lags == size of timeSeries"; 
    EXPECT_THROW(timeSeriesTransformation(origData.first, origData.first.size() + 1), std::invalid_argument) << "[ TEST_MSG ]: lags > timeSeries size"; 
}

TEST(testSplitData, testCorrectData) {
    MatrixXd X{ {1, 2, 3}, {4, 5, 6}, {7, 8, 9} };
    VectorXd y{{10, 11, 12}};
    GTEST_EXPECT_NO_DEATH({
        auto splited_data = splitData(X, y, 0.33);

        ASSERT_TRUE(splited_data.xTrain == X({ 0, 1 }, Eigen::all));
        ASSERT_TRUE(splited_data.xTest == X({ 2 }, Eigen::all));
        ASSERT_TRUE(splited_data.yTrain == y({ 0, 1 }));
        ASSERT_TRUE(splited_data.yTest == y({ 2 }));
    }) << "[ TEST_MSG ]: death splitData";
    if (HasFailure())
        TestGmdhModel::SKIP_FIXTURES = true;
}

TEST(testSplitDataWithShuffle, testCorrectData) {
    MatrixXd X{ {1, 2, 3}, {4, 5, 6}, {7, 8, 9} };
    VectorXd y{{10, 11, 12}};
    GTEST_EXPECT_NO_DEATH({
        auto splited_data = splitData(X, y, 0.66, true, 1); 
        auto splited_data2 = splitData(X, y, 0.66, true, 1);

        ASSERT_TRUE(splited_data.xTrain == splited_data2.xTrain);
        ASSERT_TRUE(splited_data.xTest == splited_data2.xTest);
        ASSERT_TRUE(splited_data.yTrain == splited_data2.yTrain);
        ASSERT_TRUE(splited_data.yTest == splited_data2.yTest);
    }) << "[ TEST_MSG ]: death splitData with shuffle";
}


TEST(testSplitData, testIncorrectData) {
    MatrixXd X{ {1, 2, 3}, {4, 5, 6}, {7, 8, 9} };
    VectorXd y{{10, 11, 12}};
    EXPECT_THROW(splitData(X({ 0, 1 }, Eigen::all), X({ 2 }, Eigen::all)({ 0, 1 }), 0.1), std::invalid_argument) << "[ TEST_MSG ]: empty test arrays because of small test_size"; 
    EXPECT_THROW(splitData(X({ 0, 1 }, Eigen::all), X({ 2 }, Eigen::all)({ 0, 1 }), 0.9), std::invalid_argument) << "[ TEST_MSG ]: empty train arrays because of large test_size"; 
    EXPECT_THROW(splitData(X({ 0, 1 }, Eigen::all), X({ 2 }, Eigen::all)({ 0 }), 0.5), std::invalid_argument) << "[ TEST_MSG ]: different X rows and y size values"; 
    EXPECT_THROW(splitData(X({ 0, 1 }, Eigen::all), X({ 2 }, Eigen::all)({ 0, 1 }), 0), std::invalid_argument) << "[ TEST_MSG ]: test_size = 0"; 
    EXPECT_THROW(splitData(X({ 0, 1 }, Eigen::all), X({ 2 }, Eigen::all)({ 0, 1 }), -1), std::invalid_argument) << "[ TEST_MSG ]: test_size < 0"; 
    EXPECT_THROW(splitData(X({ 0, 1 }, Eigen::all), X({ 2 }, Eigen::all)({ 0, 1 }), 1), std::invalid_argument) << "[ TEST_MSG ]: test_size = 1"; 
    EXPECT_THROW(splitData(X({ 0, 1 }, Eigen::all), X({ 2 }, Eigen::all)({ 0, 1 }), 10), std::invalid_argument) << "[ TEST_MSG ]: test_size > 1"; 
}

TEST(testValidateInputData, testCorrectData) {
    double testSize = 0.5, limit = 0.3;
    int pAverage = 1, threads = 2, verbose = 0, kBest = 4;
    EXPECT_EQ(validateInputData(&testSize, &pAverage, &threads, &verbose, &limit, &kBest), 0);
    threads = 0, verbose = 1;
    EXPECT_EQ(validateInputData(&testSize, &pAverage, &threads, &verbose, &limit, &kBest), 0);
    threads = 100, verbose = 6;
    EXPECT_EQ(validateInputData(&testSize, &pAverage, &threads, &verbose, &limit, &kBest), 0);
    threads = -1, verbose = -1;
    EXPECT_EQ(validateInputData(&testSize, &pAverage, &threads, &verbose, &limit, &kBest), 0);
}

TEST(testValidateInputData, testIncorrectData) {
    double testSize = 0.5, limit = 0.3;
    int pAverage = 1, threads = 2, verbose = 0, kBest = 4;

    testSize = 0;
    EXPECT_THROW(validateInputData(&testSize, &pAverage, &threads, &verbose, &limit, &kBest), std::invalid_argument);
    testSize = 1;
    EXPECT_THROW(validateInputData(&testSize, &pAverage, &threads, &verbose, &limit, &kBest), std::invalid_argument);
    testSize = 0.5;

    pAverage = 0;
    EXPECT_THROW(validateInputData(&testSize, &pAverage, &threads, &verbose, &limit, &kBest), std::invalid_argument);
    pAverage = 1;

    limit = -1;
    EXPECT_THROW(validateInputData(&testSize, &pAverage, &threads, &verbose, &limit, &kBest), std::invalid_argument);
    limit = 0;

    kBest = 0;
    EXPECT_THROW(validateInputData(&testSize, &pAverage, &threads, &verbose, &limit, &kBest), std::invalid_argument);
}

TEST(testParallelCriterion, testConstructorIncorrectData) {
    EXPECT_THROW(ParallelCriterion criterion(CriterionType::regularity, CriterionType::stability, -1), std::invalid_argument) << "[ TEST_MSG ]: alpha < 0"; 
    EXPECT_THROW(ParallelCriterion criterion(CriterionType::regularity, CriterionType::stability, 0), std::invalid_argument) << "[ TEST_MSG ]: alpha == 0"; 
    EXPECT_THROW(ParallelCriterion criterion(CriterionType::regularity, CriterionType::stability, 1), std::invalid_argument) << "[ TEST_MSG ]: alpha == 1"; 
    EXPECT_THROW(ParallelCriterion criterion(CriterionType::regularity, CriterionType::stability, 10), std::invalid_argument) << "[ TEST_MSG ]: alpha > 1"; 
}

